DatadogCore/Sources/Datadog.swift

```diff
@@ -169,6 +163,27 @@ public struct Datadog {
             self.encryption = encryption
             self.serverDateProvider = serverDateProvider ?? DatadogNTPDateProvider()
         }
+
+        // MARK: - Internal
+
+        /// Obtains OS directory where SDK creates its root folder.
+        /// All instances of the SDK use the same root folder, but each creates its own subdirectory.
+        internal var systemDirectory: () throws -> Directory = { try Directory.cache() }
```

Comment:
`Directory` is more of a convenient struct than a data object, in this case an optional `URL?` would be more appropriate.

```swift
internal var directory: URL? = nil

...

try CoreDirectory(
    in: Directory.cache(),
    in: configuration.directory.map { Directory(url: $0) } ?? try Directory.cache(),
    instancenName: instanceName,
    site: configuration.site
)
```

Reply:
Why not? It's the same as the `httpClientFactory` but in a more explicit form :)

The goal is testability and for that we need a default path and overrides for tests, we can use various solutions with more or less impacts. The closure/factory pattern is one of them but it differs from what we currently have with `additionalConfiguration`, `bundle`, or `processInfo` where the config is resolved in the `initialize` method only.

Now, it's not a big deal because it's all internal, but it's not consistent with what we have. Also, I find the following clearer than closure parameters:
```swift
/// overrides the data directory.
internal var directory: URL? = nil
/// overrides the uploader HTTP client.
internal var httpClient: HTTPClient? = nil
```

Now it's my humble opinion and I don't want to block this PR if you think otherwise.

Reply:
I see the argument on line-readability 👍, but the impacts are more than this IMO. With lack of agreed guideline on this, let's keep it as it is in this PR and discuss this topic in dev chat (I added it to the agenda).